api_definition_test.go

```diff
@@ -437,3 +435,71 @@ func TestRoundRobin(t *testing.T) {
 		}
 	}
 }
+
+func setupKeepalive(conn net.Conn) error {
+	tcpConn := conn.(*net.TCPConn)
+	if err := tcpConn.SetKeepAlive(true); err != nil {
+		return err
+	}
+	if err := tcpConn.SetKeepAlivePeriod(30 * time.Second); err != nil {
+		return err
+	}
+	return nil
+}
+
+type customListener struct {
+	L net.Listener
+}
+
+func (ln *customListener) Init(addr string) (err error) {
+	ln.L, err = net.Listen("tcp", addr)
+	return
+}
+
+func (ln *customListener) ListenAddr() net.Addr {
+	if ln.L != nil {
+		return ln.L.Addr()
+	}
+	return nil
+}
+
+func (ln *customListener) Accept() (conn io.ReadWriteCloser, clientAddr string, err error) {
+	c, err := ln.L.Accept()
+
+	if err != nil {
+		return
+	}
+
+	if err = setupKeepalive(c); err != nil {
+		c.Close()
+		return
+	}
+
+	handshake := make([]byte, 8)
+	_, err = c.Read(handshake)
+
+	// Let gorpc handle it
+	if err != nil {
+		return
+	}
```

Comment:
From the doc:

> Read reads up to len(p) bytes into p. It returns the number of bytes read (0 <= n <= len(p)) and any error encountered.

So you could get `n, err == 4, nil` for example.

I believe you want `io.ReadFull`, which will make sure `n == 8`, even if multiple `Read` calls are necessary:

> ReadFull reads exactly len(buf) bytes from r into buf.